doc/memory.rst

```diff
@@ -345,6 +346,14 @@ Gotchas
         def method(self, ...):
 	    pass
 
+  Note that the first argument (`self`) will **not** be checked as
+  part of the caching. In other words, if you have two different
+  instances `foo1`, `foo2` which are both of type `Foo` and you call
+  `foo2.method` after `foo1.method` with the same arguments then the
+  second call will return the previously cached result, even though
+  `foo1` and `foo2` are different objects.
```

Comment:
@lesteve You are completely right, my phrasing is unfortunate. I'll try to re-work it to remain closer to the truth of what's really happening under the hood.

The main "target audience" I had in mind for my explanation are users who don't necessarily know (and more importantly, don't need to care) about this subtle distinction but are simply trying to use caching for a class method. For those it is enough to know (i) how to achieve it, and (ii) that `self` is ignored for the purposes of the caching. Why exactly this is the case doesn't usually matter too much for them.

I agree that adding one or two specific examples would be very useful to understand the details.

From a user's point of view I'd personally recommend against saying that the `@mem.cache` decorator should "only be used on staticmethods or methods that don't use `self`". While technically this is true, I'm almost certain that especially the latter part will confuse less advanced users, because surely the definition
```
def method(self, ...):
```
uses the variable `self`? Not sure how many users actually know about the nitty-gritty details about how binding of the first argument works on class methods in Python -- especially at which point the binding actually happens when your code is executed (e.g., when an instance of the class is instantiated, or only when the method is actually called on that instance?), because this is the bit that matters to understand this gotcha about the `@mem.cache` decorator.

But again, I'd argue that most users don't need to fully understand these details but can still happily use this feature, if they are shown how to do it and what to look out for (which is why a couple of well-chosen examples would help a lot).